---
title: "GroupProject_codeRMD"
author: "Bhavana"
date: "December 5, 2017"
output:
  pdf_document: default
  html_document: default
---
###Code for group project
###Subproblem: Translating RNAseq files

```
##Translating RNAseq files into aminoacids
##
##Read the aa sequence file - codonmap.txt file
codondata=scan(file = "codonmap.txt", what = character(), sep = "\n")
##
#Assign empty files of character vector format to extract open reading frames into them
match1=character(length(294))
match2=character(length(400))
match3=character(length(912))
match4=character(length(880))
##
##
#Read the nucleotide fasta files control 1&2 and obese 1&2
control1=scan(file = "control1.fasta.txt", what = character(), sep = "\n")
control2=scan(file = "control2.fasta.txt", what = character(), sep = "\n")
obese1=scan(file = "obese1.fasta.txt", what = character(), sep = "\n")
obese2=scan(file = "obese2.fasta.txt", what = character(), sep = "\n")
##
##
#Load stringr/i and seqinr packages
library(stringr)
library(stringi)
library(seqinr)
##
##
#Formulating regex for identifying open reading frame
ATG([ATCG]{3})+?(TAA|TAG|TGA)
##
##
#forloop for control1 seq file to extract reading frame
for (Line in 1:length(control1)){
#Operate only on lines that do not include >, skips header lines
#Note that str_detect returns a logical T/F
if (!str_detect(control1[Line],">")==TRUE){
#Find the ORF in each sequence line
match1 = str_extract(control1[Line],"ATG([ATCG]{3})+?(TAA|TAG|TGA)")
#Print the ORF to standard out
print(match1)
  }
}
##
##
#use translate function to translate nucleotides into amino acids
control1aa=(translate(s2c(match1)))
#convert vector of aa into a dataframe      
seqc1<-as.data.frame(control1aa)
#print the output aminoacid sequence
seqc1
##
##
#forloop for control2 seq file to extract reading frames
for (Line in 1:length(control2)){
  #Operate only on lines that do not include >, skips header lines
  #Note that str_detect returns a logical T/F
  if (!str_detect(control2[Line],">")==TRUE){
    #Find the ORF in each sequence line
    match2 = str_extract(control2[Line],"ATG([ATCG]{3})+?(TAA|TAG|TGA)")
    #Print the ORF to standard out
    print(match2)
  }
}
##
##
#use translate function to translate nucleotides into aa
control2aa=(translate(s2c(match2)))
#convert vector of aa into a dataframe      
seqc2<-as.data.frame(control2aa)
##
##
#forloop for obese1 seq file orf extract
for (Line in 1:length(obese1)){
  #Operate only on lines that do not include >, skips header lines
  #Note that str_detect returns a logical T/F
  if (!str_detect(obese1[Line],">")==TRUE){
    #Find the ORF in each sequence line
    match3 = str_extract(obese1[Line],"ATG([ATCG]{3})+(TAA|TAG|TGA)?")
    #Print the ORF to standard out
    print(match3)
  }
}
##
##
#use translate function to translate nucleotides into aa
obese1aa=(translate(s2c(match3)))
#convert vector of aa into a dataframe      
seqo1<-as.data.frame(obese1aa)
##
##
#forloop for obese2 seq file orf extract
for (Line in 1:length(obese2)){
  #Operate only on lines that do not include >, skips header lines
  #Note that str_detect returns a logical T/F
  if (!str_detect(obese2[Line],">")==TRUE){
    #Find the ORF in each sequence line
    match4 = str_extract(obese2[Line],"ATG([ATCG]{3})+(TAA|TAG|TGA)?")
    #Print the ORF to standard out
    print(match4)
  }
}
##
##
#use translate function to translate nucleotides into aa
obese2aa=(translate(s2c(match4)))
#convert vector of aa into a dataframe      
seqo2<-as.data.frame(obese2aa)

```
###Subproblem: Building HMM profiles and performing hmm searches


```
##Build a Hidden Markov Model for each of the 6 transcript proteins and search
##the 4 translated "RNAseq files."
##
##   
## align each of the reference files with muscle into .align file
##NOTE:all the files must be in the same place for convenience or else need to specify
##the path to each file. Here all my files are in the current working directory
##
##
# ./muscle3.8.31_i86win32.exe -in Transcript10_Synpr.fasta -out Transcript10.align
# ./muscle3.8.31_i86win32.exe -in Transcript9_Lhx2.fasta -out Transcript9.align
# ./muscle3.8.31_i86win32.exe -in Transcript8_Atp12a.fasta -out Transcript8.align
# ./muscle3.8.31_i86win32.exe -in Transcript6_Ptpn5.fasta -out Transcript6.align
# ./muscle3.8.31_i86win32.exe -in Transcript2_Slc7a12.fasta -out Transcript2.align
# ./muscle3.8.31_i86win32.exe -in Transcript1_Gsta2.fasta -out Transcript1.align
## 
##
##forloop muscle alignment
#for file in ./*.fasta
#do
#./muscle3.8.31_i86win32.exe -in $file -out $file.align
#done
##
##
##for loop HMM profile Build
#for file in ./*.align
#do
#./hmmbuild.exe $file.hmm $file
#done

## 
##run hmmsearch with the --tblout option and use the
##index variable to specify a unique output file
##do the hmmsearch 6 times, one per HMM profile
# 
#
#for loop for hmmsearch of four RNAseq files and 6 transcript files
#for file in ./*.fasta
#do
#./hmmsearch.exe --tblout $file.hits Transcript1.align.hmm $file
#done
#for file in ./*.fasta
#do
#./hmmsearch.exe --tblout $file.hits Transcript2.align.hmm $file
#donefor file in ./*.fasta
#do
#./hmmsearch.exe --tblout $file.hits Transcript9.align.hmm $file
#donefor file in ./*.fasta
# do
#./hmmsearch.exe --tblout $file.hits Transcript6.align.hmm $file
#donefor file in ./*.fasta
#do
#./hmmsearch.exe --tblout $file.hits Transcript8.align.hmm $file
#done
#for file in ./*.fasta
#do
#./hmmsearch.exe --tblout $file.hits Transcript10.align.hmm $file
#done
```







```
```

