---
title: "GroupProject_codeRMD"
author: "Bhavana"
date: "December 5, 2017"
output:
  pdf_document: default
  html_document: default
---
###Code for group project

```
###Subproblem: Building HMM profiles and performing hmm searches


```
##Build a Hidden Markov Model for each of the 6 transcript proteins and search
##the 4 translated "RNAseq files."
##
##unix commands   
## align each of the reference files with muscle into .align file
##NOTE:all the files must be in the same place for convenience or else need to specify
##the path to each file. Here all my files are in the current working directory
##
##
# ./muscle3.8.31_i86win32.exe -in Transcript10_Synpr.fasta -out Transcript10.align
# ./muscle3.8.31_i86win32.exe -in Transcript9_Lhx2.fasta -out Transcript9.align
# ./muscle3.8.31_i86win32.exe -in Transcript8_Atp12a.fasta -out Transcript8.align
# ./muscle3.8.31_i86win32.exe -in Transcript6_Ptpn5.fasta -out Transcript6.align
# ./muscle3.8.31_i86win32.exe -in Transcript2_Slc7a12.fasta -out Transcript2.align
# ./muscle3.8.31_i86win32.exe -in Transcript1_Gsta2.fasta -out Transcript1.align
## 
##
##forloop muscle alignment
#for file in ./*.fasta
#do
#./muscle3.8.31_i86win32.exe -in $file -out $file.align
#done
##
##
##for loop HMM profile Build
#for file in ./*.align
#do
#./hmmbuild.exe $file.hmm $file
#done

## 
##run hmmsearch with the --tblout option and use the
##index variable to specify a unique output file
##do the hmmsearch 6 times, one per HMM profile
# 
#
#
```
###Subproblem: Translating RNAseq files

```
##Translating RNAseq files into aminoacid sequences
##Using regex and translate function from seqinr 
##
#Assign empty character vectors for file to extract open reading frame into them
match1=character(length(control1)/2)
match2=character(length(control2)/2)
match3=character(length(obese1)/2)
match4=character(length(obese2)/2)
##
##
#Read the nucleotide fasta files control 1&2 and obese 1&2
control1=scan(file = "control1.fasta.txt", what = character(), sep = "\n")
control2=scan(file = "control2.fasta.txt", what = character(), sep = "\n")
obese1=scan(file = "obese1.fasta.txt", what = character(), sep = "\n")
obese2=scan(file = "obese2.fasta.txt", what = character(), sep = "\n")
##
##
#Load stringr/i and seqinr packages
library(stringr)
library(stringi)
library(seqinr)
##
##
##
##
#forloop to extraxt openreading frame of control1 seq file
for (Line in 1:length(control1)){
#Operate only on lines that do not include >, skips header lines
#Note that str_detect returns a logical T/F
if (!str_detect(control1[Line],">")==TRUE){
#Find the ORF in each sequence line
match1[Line/2] = str_match(control1[Line],"([ATCG]{3})*?(ATG([ATCG]{3})+?(TAA|TAG|TGA))")[,3]
#Print the ORF to standard out
print(match1)
  }
}
##
#create an empty vector of length equal to match1 each for tranlated amino acid sequences before and after formating
control1aa=character(length(match1))
control1protein=c(length(match1))
#use translate function to translate nucleotides into aa through a for loop
for(Line in 1:length(match1)){
#translate sequence of each line from the match1 in a loop
  control1aa=(translate(s2c(match1[Line])))
#print the translated sequence to standard out
  print(control1aa)
#format the output by removing quotes from each amino acid
  control1protein[Line]=paste(control1aa, sep="", collapse="")
  print(control1protein)
#write the sequence of aminoacids in a fasta file format
  write.fasta(sequences = as.list(control1protein), names = rep("control1", times=length(control1protein)),file.out = "control1protein.fasta")
}
##
##
##Repeat the same for all the original files(control2,obese1,and obese2)
##
##
#forloop for control2 seq file orf extract
for (Line in 1:length(control2)){
#Operate only on lines that do not include >, skips header lines
#Note that str_detect returns a logical T/F
if (!str_detect(control2[Line],">")==TRUE){
#Find the ORF in each sequence line
match2[Line/2] = str_match(control2[Line],"([ATCG]{3})*?(ATG([ATCG]{3})+?(TAA|TAG|TGA))")[,3]
#Print the ORF to standard out
print(match2)
}
}
##
#create an empty vector of length equal to match2 each for storing translated aminoacid before and after formating
control2aa=character(length(match2))
control2protein=c(length(match2))
#use translate function to translate nucleotides into aa through a for loop
for(Line in 1:length(match2)){
#translate sequence of each line from the match1 in a loop
  control2aa=(translate(s2c(match2[Line])))
#print the translated sequence to standard out
  print(control2aa)
#format the output by removing quotes from each amino acid
  control2protein[Line]=paste(control2aa, sep="", collapse="")
#print the formated translated sequence to standard out
  print(control2protein)
#write the sequence of aminoacids in a fasta file format
  write.fasta(sequences = as.list(control2protein), names = rep("control2", times=length(control2protein)),file.out = "control2protein.fasta")
}
##
##
#forloop for obese1 seq file orf extract
for (Line in 1:length(obese1)){
  #Operate only on lines that do not include >, skips header lines
  #Note that str_detect returns a logical T/F
  if (!str_detect(obese1[Line],">")==TRUE){
    #Find the ORF in each sequence line
    match3[Line/2] = str_match(obese1[Line],"([ATCG]{3})*?(ATG([ATCG]{3})+?(TAA|TAG|TGA))")[,3]
    #Print the ORF to standard out
    print(match3)
  }
}
##
##
##
#create an empty vector of length equal to match3 each 
#for storing translated aminoacid before and after formating
obese1aa=character(length(match3))
obese1protein=c(length(match3))
##
##
#use translate function to translate nucleotides into aa through a for loop
for(Line in 1:length(match3)){
#translate sequence of each line from the match1 in a loop
obese1aa=(translate(s2c(match3[Line])))
#print the translated sequence to standard out
print(obese1aa)
#format the output by removing quotes from each amino acid
obese1protein[Line]=paste(obese1aa, sep="", collapse="")
#print the formated translated sequence to standard out
print(obese1protein)
#write the sequence of aminoacids in a fasta file format
  write.fasta(sequences = as.list(obese1protein), names = rep("obese1", times=length(obese1protein)),file.out = "obese1protein.fasta")
  }
##
##
##
#forloop for obese2 seq file orf extract
for (Line in 1:length(obese2)){
#Operate only on lines that do not include >, skips header lines
#Note that str_detect returns a logical T/F
if (!str_detect(obese2[Line],">")==TRUE){
#Find the ORF in each sequence line
match4[Line/2] = str_match(obese2[Line],"([ATCG]{3})*?(ATG([ATCG]{3})+?(TAA|TAG|TGA))")[,3]
    #Print the ORF to standard out
    print(match4)
}
  }
##
##
#create an empty vector of length equal to match4 each 
#for storing translated aminoacid before and after formating
obese2aa=character(length(match4))
obese2protein=c(length(match4))
##
##
#use translate function of seqinr package to translate nucleotides into aa
for(Line in 1:length(match4)){
#translate sequence of each line from the match1 in a loop
obese2aa=(translate(s2c(match4[Line])))
#print the translated sequence to standard out
print(obese2aa)
#format the output by removing quotes from each amino acid
obese2protein[Line]=paste(obese2aa, sep="", collapse="")
#print the formated translated sequence to standard out
print(obese2protein)
#write the sequence of aminoacids in a fasta file format
write.fasta(sequences = as.list(obese2protein), names = rep("obese2", times=length(obese2protein)),file.out = "obese2protein.fasta")
  }
##further exploration-1
##Cygwin script for table of tophits
$ cat Transcript1_hitstable.txt.csv | head -1 >> hitstab.txt
$ cat Transcript2_hitstable.txt.csv | head -1 >> hitstab.txt
$ cat Transcript6_hitstable.txt.csv | head -1 >> hitstab.txt
$ cat Transcript8_hitstable.txt.csv | head -1 >> hitstab.txt
$ cat Transcript9_hitstable.txt.csv | head -1 >> hitstab.txt
$ cat Transcript10_hitstable.txt.csv | head -1 >> hitstab.txt
##script for discontinuous megablast
$ cat Transcript1_hittable.csv | head -1 >> dismegahitstab.txt
$ cat Transcript2_hittable.csv | head -1 >> dismegahitstab.txt
$ cat Transcript6_hittable.csv | head -1 >> dismegahitstab.txt
$ cat Transcript8_hittable.csv | head -1 >> dismegahitstab.txt
$ cat Transcript9_hittable.csv | head -1 >> dismegahitstab.txt
$ cat Transcript10_hittable.csv | head -1 >> dismegahitstab.txt
##table of blast hits- discontinuous megablast
Transcript_1,gi|118130920|ref|NM_008182.3|,Transcript_1,NM_008182.3,100.000,854,0,0,1,854,2,855,0.0,1541
Transcript_2,gi|142366209|ref|NM_080852.2|,Transcript_2,NM_080852.2,100.000,1752,0,0,1,1752,3,1754,0.0,3160
Transcript_6,gi|254675196|ref|NM_013643.2|,Transcript_6,NM_013643.2,100.000,2989,0,0,1,2989,146,3134,0.0,5391
Transcript_8,gi|157168325|ref|NM_138652.2|,Transcript_8,NM_138652.2,100.000,3954,0,0,1,3954,1,3954,0.0,7131
Transcript_9,gi|594591606|ref|NM_010710.4|,Transcript_9,NM_010710.4,100.000,2294,0,0,1,2294,2,2295,0.0,4138
Transcript_10,gi|253314531|ref|NM_028052.4|,Transcript_10,NM_028052.4,100.000,2550,0,0,1,2550,2,2551,0.0,4599
##blastn
$ cat Transcript1_hittable.csv | head -1 >> blastnhitstab.txt
$ cat Transcript2_hittable.csv | head -1 >> blastnhitstab.txt
$ cat Transcript6_hittable.csv | head -1 >> blastnhitstab.txt
$ cat Transcript8_hittable.csv | head -1 >> blastnhitstab.txt
$ cat Transcript9_hittable.csv | head -1 >> blastnhitstab.txt
$ cat Transcript10_hittable.csv | head -1 >> blastnhitstab.txt
##table of blast hits - blastn quality
Transcript_1,gi|118130920|ref|NM_008182.3|,Transcript_1,NM_008182.3,100.000,854,0,0,1,854,2,855,0.0,1541
Transcript_2,gi|142366209|ref|NM_080852.2|,Transcript_2,NM_080852.2,100.000,1752,0,0,1,1752,3,1754,0.0,3160
Transcript_6,gi|254675196|ref|NM_013643.2|,Transcript_6,NM_013643.2,100.000,2989,0,0,1,2989,146,3134,0.0,5391
Transcript_8,gi|157168325|ref|NM_138652.2|,Transcript_8,NM_138652.2,100.000,3954,0,0,1,3954,1,3954,0.0,7131
Transcript_9,gi|594591606|ref|NM_010710.4|,Transcript_9,NM_010710.4,100.000,2294,0,0,1,2294,2,2295,0.0,4138
Transcript_10,gi|253314531|ref|NM_028052.4|,Transcript_10,NM_028052.4,100.000,2550,0,0,1,2550,2,2551,0.0,4599





```
```

