---
title: "Pseudocode for Searching for a Coding Sequence"
author: "Shane Davitt"
date: "November 29, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
This is a plan for taking a fasta nucleotide sequence (RNA) file as input and searching for a protein-coding strand. Note that this code could also be made into a script, with parameters to replace file names and such.

There are three main steps:

1) Load the data from a file into an object. 
--Could use read.table or similar functions, but they
often leave the ">Header" within the dataframe, and   thus require processing before the sequence can be    searched. The scan function has similar problems. "read.fasta" with as.string=TRUE removes these headers and produces one long string, but it appears to be a different data type and not-immediately subsettable, which could be a problem if using a for loop.

2) Read the sequence by 3's for a coding sequence and match to a codonmap. 
--The instructions make it seem as if the only reading frame with a correct start codon takes the first nucleotide of the file as the begining of the first triplet (should double check). Thus we need to read the sequence by triplets. Using a for loop may be what the instructors want, but may be clunky to select sequential triplets. Could possibly use integer multipliers of the index variable to select certain positions in each triplet and then later merge together for sorting, but how exactly to do this is as yet unclear.
The codonmap should then be read into a datatframe. Once split up, each triplet can be searched against the codonmap triplets and matched to a single letter. Then regex could be used to find a sequence which starts with a start codon letter and ends with a stop codon letter. This is the coding sequence. 

3) The coding sequence should then be passed and saved into a new fasta file (as per instructions).